### PR TITLE
Make __call__ work properly for Graph

### PR DESCRIPTION
**Pull request by neggert**
Previously, `Graph` inherited `__call__` from `Layer`, which doesn't work as expected. I added a bunch of tests that previously failed, but the basic problem is that this example throws an `UnusedInputError`

```{python}
graph = Graph()
graph.add_input("input", input_shape=(10, ))
graph.add_node(Dense(10), input="input", name="output", create_output=True)

inputA = K.placeholder(ndim=2)
outputA = graph(inputA)
fA = K.function([inputA], outputA)
```

It also breaks Siamese graphs, since the Siamese implementation relies on `__call__`. This example also throws an `UnusedInputError`

```{python}
graph = Graph()
graph.add_input("input", input_shape=(10, ))
graph.add_node(Dense(10), input="input", name="output", create_output=True)

model = Graph()
model.add_input("inputA", input_shape=(10, ))
model.add_input("inputB", input_shape=(10, ))
model.add_shared_node(name="output", layer=graph, inputs=["inputA", "inputB"],
                      merge_mode="cos", dot_axes=1, create_output=True)
model.compile(loss={"output": "binary_crossentropy"}, optimizer="sgd")
```

This ended up requiring a number of improvements

### Short version

* Changed implementation of `Layer.__call__`
* Remove overridden `Sequential.__call__` (new inherited implementation works)
* `Layer.set_previous` takes a new parameter to disable rebuild
* Cases that previously set the `previous` attribute manually now use `set_previous`
* Added `clear_previous` methods
* Added `Graph.input_shape` property
* Implemented `Graph.__call__` for multiple inputs
* Implemented `Graph.cache_enabled` property, which passes the property to graph nodes
* Added new tests for `__call__` on `Graph` models

### Long Version
I changed the implementation of `Layer.__call__`, so that instead of messing with `get_input`, it creates a dummy layer and uses `set_previous` to set the bottom layer's input. This implementation works for normal layers as well as `Sequential` and `Graph`s with only 1 input. The reason that this approach works better is that it respects layers with custom `set_previous` methods (like Graph)

To make this work, I ended up adding an option to `Layer.set_previous` to disable the rebuild, so that a layer's input can be changed without re-generating its trainable parameters. This allows us to get rid of a few places where `previous` attributes are set manually, which causes problems for layers that have custom `set_previous` implementations (like Graph).

I also needed to add `clear_previous` methods all around, add an `input_shape` method to `Graph`, and add an implementation of the `Graph.cache_enabled` that passes the property to all the graph nodes.

Finally, I added an override implementation of `Graph.__call__` for cases with multiple inputs. I figured as long as I'm fixing it, I may as well make it feature-complete.

There are new tests to cover all of the things I fixed. Since there were a few preparatory steps before the main change, I split this into a few commits, but I'd be happy to squash them if you'd prefer.

### CODE REVIEW

**Comment by neggert** [comment]
Hmm, those tests passed on my machine. I'll take a look.

**Comment by fchollet** [comment]
I suppose you did not run the tests with TensorFlow.

**Comment by neggert** [comment]
Yeah, I thought it was automatically running both Theano and TF, since it looks like the tests need TF installed even with the Theano backend. Anyway, I reproduced it and have a fix. Checking to make sure everything else checks out with TF, then I'll push the commits.

**Comment by neggert** [comment]
Should be good to go now.

**Comment by fchollet** [comment]
The addition of the `build` argument to `set_previous`, and the addition of `clear_previous`, is not necessary and adds undesirable complexity to the code. You can simply follow the pattern in use for the existing implementation of `__call__` (overriding `get_input`).

**Comment by neggert** [comment]
The issue is that, for containers (`Graph` and `Sequential`), changing `get_input` doesn't work. Currently, `Sequential` handles this by overriding `__call__`, and `Graph` doesn't handle it at all.

So the alternative solution is for all containers to override `__call__`. This strikes me as being more brittle than my solution.

The other advantage of adding a `build` argument to `set_previous` is that it lets you get rid of the cases where `previous` is set manually because we don't want to trigger a build (`Sequential` and `AutoEncoder`). Setting `previous` manually is brittle and doesn't work for e.g. `Graph`s. 

I suppose one other option would be to completely decouple `previous` and `build`. I think this would just require adding a call to `build` in `Sequential.add` and `Graph.add_node`.

Anyway, I'm happy to implement whatever you think is best.

**Comment by fchollet** [comment]
I think the `build` argument makes sense. However it is not immediately clear what it does unless one is familiar with the Keras internals. What I would do:

- rename it `reset_weights`, which makes its effect clear. If set to `False`, a check should be made that the shape of the current input(s) matches the shape of the weights (i.e. is compatible with the shape of the previously connected input), and otherwise an error should be raised. This is necessary to prevent user errors where a layer would be connected to an incompatible input.
- get rid of `clear_previous`. Since a layer cannot exist without either a previous layer or being registered as an input, `clear_previous` is not conceptually sound.

**Comment by neggert** [comment]
Sounds good. I'll get on it.

**Comment by neggert** [comment]
@fchollet How about `init_weights` instead of `reset_weights`, since it's used to initially create the weights as well?

**Comment by neggert** [comment]
After looking at this a bit, I think it's going to be impossible to implement `__call__` with a dummy layer without a `clear_previous` method. I can delete the `previous` attribute manually, but that won't work for containers.

Here's completely different idea for getting this working: Let's make it so that overriding `get_input` *does* work for containers. I think what this would involve is changing `Sequential` and `Graph` so that the bottom layers have `previous` pointing to the container. Then the implementation of `get_input` for containers will look a lot like the one for a normal layer. 

I haven't fully figured out how to do this yet, but I think the call stack would look something like this:

```
Sequential.get_output
	-> Layer.get_output
		-> Layer.get_input
			->Sequential.get_input
				-> Sequential.previous.get_output
				...
```

**Comment by neggert** [comment]
Bah! What I talked about above would require a fairly big overhaul of `Graph` that I'm not really up for.

How about I add a check in `clear_previous` so that it throws an error if `input` isn't set? It *is* possible to have *both* `previous` and `input` set. In that case, `previous` takes precedence, which is what my `__call__` implementation relies on.

**Comment by fchollet** [comment]
> How about I add a check in clear_previous so that it throws an error if input isn't set?

Alright.

**Comment by neggert** [comment]
@fchollet Ready to go hopefully.

File: keras/layers/containers.py [hunk 1]
```diff
@@ -251,22 +272,35 @@ def reset_states(self):
             if hasattr(l, 'reset_states') and getattr(l, 'stateful', False):
                 l.reset_states()
 
-    def set_previous(self, layer, connection_map={}):
+    def set_previous(self, layer, connection_map={}, init_weights=True):
```
Comment (fchollet) [inline keras/layers/containers.py]: There is already a variable named `initial_weights`. To avoid confusion, please use `reset_weights`.

**Comment by fchollet** [comment]
Otherwise, this looks good to me. Please squash your commits into a single PR commit.

**Comment by neggert** [comment]
Done. Not sure why Travis isn't running...

**Comment by fchollet** [comment]
- `init_weights` was not renamed everywhere
- it doesn't seem to me that any checks are made to prevent users from calling `set_previous` with `reset_weights=False` on an incompatible previous layer. Can you clarify?

**Comment by neggert** [comment]
Sorry, I misunderstood the check you were asking for. Re-reading your earlier comment made it clear. I added a check that compares `layer.output_shape` to `self.input_shape`. Also fixed the `init_weights` change I missed.

**Comment by fchollet** [comment]
Looks good to me. Thank you for the contribution!
